source/common/formatter/substitution_formatter.cc

```diff
@@ -1918,6 +1925,69 @@ UpstreamHostMetadataFormatter::UpstreamHostMetadataFormatter(const std::string&
                           return host->metadata().get();
                         }) {}
 
+ProxyProtocolTlvsFormatter::ProxyProtocolTlvsFormatter(const std::string& tlv_type) {
+  ASSERT(!tlv_type.empty());
+  // Specified tlv_type must be parsable as an int.
+  try {
+    tlv_type_ = std::stoi(tlv_type);
+  } catch (const EnvoyException& ex) {
+    ENVOY_LOG_MISC(error, "Invalid parameter provided for PROXY_PROTOCOL_TLVS header: {}", tlv_type);
+    // Arbitrary int that won't be used for proxy procotol TLV type. 
+    tlv_type_ = -1; 
+  }
+}
+
+// Parses the proxy protocol TLVs stored in the FilterState with the key "envoy.network.proxy_protocol_options" 
+// and returns a function suitable for accessing the specified metadata from a StreamInfo::StreamInfo. 
+absl::optional<std::string> ProxyProtocolTlvsFormatter::format(const Http::RequestHeaderMap&,
+                                                        const Http::ResponseHeaderMap&,
+                                                        const Http::ResponseTrailerMap&,
+                                                        const StreamInfo::StreamInfo& stream_info,
+                                                        absl::string_view) const {
+  const Envoy::StreamInfo::FilterState& filter_state = stream_info.filterState();
+  auto typed_state = filter_state.getDataReadOnly<Network::ProxyProtocolFilterState>(Network::ProxyProtocolFilterState::key());
```

Comment:
Interesting, many other usages that I see of `getDataReadOnly()` assigns the returned value to a variable of type `auto`, not `auto&` ([example](https://github.com/10gen/envoy-serverless/blob/aa484fbfcfd6aa19544a860d45e9794ff0f2ee73/source/common/network/transport_socket_options_impl.cc#L66-L96), [example](https://github.com/10gen/envoy-serverless/blob/aa484fbfcfd6aa19544a860d45e9794ff0f2ee73/source/common/http/hash_policy.cc#L166)). I also see this example where the returned value is saved as `const auto*` ([link](https://github.com/10gen/envoy-serverless/blob/aa484fbfcfd6aa19544a860d45e9794ff0f2ee73/source/common/tcp_proxy/tcp_proxy.cc#L147)). Is there any benefit to storing it as just `auto` or `const auto*`? 

Also, when I tried compiling with the type set to `auto&`, I saw the following error: 
```
source/common/formatter/substitution_formatter.cc:1947:9: error: non-const lvalue reference to type 'const Envoy::Network::ProxyProtocolFilterState *' cannot bind to a temporary of type 'const Envoy::Network::ProxyProtocolFilterState *'
  auto& typed_state = filter_state.getDataReadOnly<Network::ProxyProtocolFilterState>(Network::ProxyProtocolFilterState::key());
        ^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
```
I believe this can be fixed by using `const auto&`. However, we could also avoid using `auto` altogether with `const Network::ProxyProtocolFilterState*` because we do know the specific type. Curious to hear your thoughts here!

Reply:
Ah, I see. Your linked code is in a newer version than mine. If it returns a pointer, then it makes sense to just use `auto` and check with `nullptr`. Sorry for the confusion.